### PR TITLE
Add Even More Knowledge part

### PR DESCRIPTION
**Pull request by thovo**
Please review and merge! Sorry for delay, some words is too technical.

### CODE REVIEW

**Comment by letientai299** [comment]
Thanks. I'm a little bit busy now. Will review and get back to you soon.

File: README-vn.md [hunk 1]
```diff
@@ -1289,57 +1289,57 @@ Bạn sẽ biết thêm nhiều ứng dụng của đồ thị trong sách của
...
-    - [ ] [2015: TensorFlow: Large-Scale Machine Learning on Heterogeneous Distributed Systems](http://download.tensorflow.org/paper/whitepaper2015.pdf )
-    - [ ] [2015: How Developers Search for Code: A Case Study](http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43835.pdf)
+    - [ ] [2014: Máy học: Lợi ích và những vấn đề cần tránh (Machine Learning: The High-Interest Credit Card of Technical Debt)](http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43146.pdf)
+    - [ ] [2015: Continuous Pipelines at Google](http://static.googleusercontent.com/media/research.google.com/en/pubs/archive/43790.pdf)
+    - [ ] [2015: Sẵn sàng cho thay đổi lớn: Xây dựng hạ tầng dữ liệu của Google dành cho quảng cáo (High-Availability at Massive Scale: Building Google’s Data Infrastructure for Ads)](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/44686.pdf)
+    - [ ] [2015: TensorFlow: Áp dụng máy học trên diện rộng với hệ thống phân tán không đồng nhất (Large-Scale Machine Learning on Heterogeneous Distributed Systems)](http://download.tensorflow.org/paper/whitepaper2015.pdf )
```
Comment (dangkhoasdc) [inline README-vn.md]: Should be: "Tensorflow: Áp dụng máy học với dữ liệu lớn trên hệ thống phân tán tính toán song song . I think it makes more sense although the translation isn't correctly word by word.

File: README-vn.md [hunk 2]
```diff
@@ -1289,57 +1289,57 @@ Bạn sẽ biết thêm nhiều ứng dụng của đồ thị trong sách của
...
+        - [ ] [2. Tìm và thay chuỗi theo Brute-Force(Brute-Force Substring Search)](https://www.youtube.com/watch?v=CcDXwIGEXYU&list=PLe-ggMe31CTdAdjXB3lIuf2maubzo9t66&index=4)
         - [ ] [3. Knuth-Morris Pratt](https://www.youtube.com/watch?v=n-7n-FDEWzc&index=3&list=PLe-ggMe31CTdAdjXB3lIuf2maubzo9t66)
         - [ ] [4. Boyer-Moore](https://www.youtube.com/watch?v=fI7Ch6pZXfM&list=PLe-ggMe31CTdAdjXB3lIuf2maubzo9t66&index=2)
         - [ ] [5. Rabin-Karp](https://www.youtube.com/watch?v=QzI0p6zDjK4&index=1&list=PLe-ggMe31CTdAdjXB3lIuf2maubzo9t66)
-    - [ ] [Search pattern in text (video)](https://www.coursera.org/learn/data-structures/lecture/tAfHI/search-pattern-in-text)
+    - [ ] [Tìm kiếm mẫu trong văn bản (Search pattern in text) (video)](https://www.coursera.org/learn/data-structures/lecture/tAfHI/search-pattern-in-text)
 
-    If you need more detail on this subject, see "String Matching" section in [Additional Detail on Some Subjects](#additional-detail-on-some-subjects)
+    Nếu bạn cần thêm thông tin chi tiết, hãy đọc qua phần "Khớp chuỗi" trong các mục đọc thêm [Đọc thêm về một số đề tài](đọc-thêm-về-một-số-đề-tài)
```
Comment (dangkhoasdc) [inline README-vn.md]: So khớp chuỗi.

File: README-vn.md (hunk 1 above)
Comment (thovo) [reply README-vn.md]: I am not sure about `tính toán song song` part, is it correct term for `Heterogeneous ` ? We should use the word with nearest meaning

Comment (dangkhoasdc) [reply README-vn.md]: Sometimes we have to change the true meaning of a word for easy understanding :smile: 
As we know, "heterogeneous" implies the computing between "cpu" and "gpu". The term seems more difficult to understand than "gpu computing".

Comment (thovo) [reply README-vn.md]: I tried to look over the paper to find out there is any explanation :smile: but I didnt find out anything useful, so if it is the nearest meaning we have, we may use it

Comment (dangkhoasdc) [reply README-vn.md]: For example, [Heterogeneous Parallel Programming](https://www.coursera.org/signature/course/hetero/973470) is a course about GPU computing :D

Comment (thovo) [reply README-vn.md]: It made me confusion, not sure about `Programming` part is the same like `Distributed System`, we may not apply the same meaning for two different objects

Comment (dangkhoasdc) [reply README-vn.md]: Actually, Tensorflow is a library for deep learning which it utilizes the power of GPU :p

Comment (thovo) [reply README-vn.md]: I am unsure, please vote @letientai299 ! :smile:

Comment (thovo) [reply README-vn.md]: As we got no reply, so I will change the translation and make new pull request by end of day

**Comment by thovo** [comment]
@dangkhoasdc and @letientai299 please review again before merge !

File: README-vn.md [hunk 3]
```diff
@@ -1130,14 +1130,218 @@ Bạn sẽ biết thêm nhiều ứng dụng của đồ thị trong sách của
...
-        - [Thiết kế một hệ thống chia sẽ ảnh (Design a picture sharing system) ](http://highscalability.com/blog/2011/12/6/instagram-architecture-14-million-users-terabytes-of-photos.html)
-        - [Thiết kế một hệ thống hổ trợ ra quyết định (Design a recommendation system)](http://ijcai13.org/files/tutorial_slides/td3.pdf)
-        - [Thiết kế một hệ thống làm ngắn URL (Design a URL-shortener system: copied from above)](http://www.hiredintech.com/system-design/the-system-design-process/)
-        - [Thiết kế một hệ thống cache (Design a cache system)](https://www.adayinthelifeof.nl/2011/02/06/memcache-internals/)
+        - [Thiết kế một hàm để trả về những từ khóa được tìm kiếm nhiều nhất trong thời gian gần nhất(Design a function to return the top k requests during past time interval)]( https://icmi.cs.ucsb.edu/research/tech_reports/reports/2005-23.pdf)
+        - [Thiết kế một hệ thống chia sẽ ảnh(Design a picture sharing system) ](http://highscalability.com/blog/2011/12/6/instagram-architecture-14-million-users-terabytes-of-photos.html)
```
Comment (dangkhoasdc) [inline README-vn.md:1137]: chia se?

File: README-vn.md [hunk 4]
```diff
@@ -1130,14 +1130,218 @@ Bạn sẽ biết thêm nhiều ứng dụng của đồ thị trong sách của
...
+        - [Thiết kế một hệ thống hổ trợ ra quyết định(Design a recommendation system)](http://ijcai13.org/files/tutorial_slides/td3.pdf)
+        - [Thiết kế một hệ thống làm ngắn URL(Design a URL-shortener system: copied from above)](http://www.hiredintech.com/system-design/the-system-design-process/)
+        - [Thiết kế một hệ thống cache(Design a cache system)](https://www.adayinthelifeof.nl/2011/02/06/memcache-internals/)
+
+## Các kiến thức thêm
```
Comment (dangkhoasdc) [inline README-vn.md:1142]: đọc thêm

File: README-vn.md [hunk 5]
```diff
@@ -1130,14 +1130,218 @@ Bạn sẽ biết thêm nhiều ứng dụng của đồ thị trong sách của
...
+        - [ ] [Đệ qui đuôi (Tail Recursion) (video)](https://www.youtube.com/watch?v=L1jjXGfxozc)
+
+- ### Lập trình hướng đối tượng
+    - [ ] [Phụ thêm: Seri về UML 2.0 (Optional: UML 2.0 Series) (video)](https://www.youtube.com/watch?v=OkC7HKtiZC0&list=PLGLfVvz_LVvQ5G-LdJ8RLqe-ndo7QITYc)
+    - [ ] Lập trình hướng đối tượng: Phát triển phần mềm sử dụng UML và Java (21 videos):
+        - Có thể bỏ qua phần này nếu bạn đã có kiến thức vững chắc về OOP
+        - [Lập trình hướng đối tượng: Phát triển phần mềm sử dụng UML và Java](https://www.youtube.com/playlist?list=PLJ9pm_Rc9HesnkwKlal_buSIHA-jTZMpO)
+    - [ ] Những điểm quan trọng nhất của OOP:
+        - [ ] [Bob Martin Những điểm quan trọng nhất của lập trình hướng đối tượng và thiết kế linh hoạt (Bob Martin SOLID Principles of Object Oriented and Agile Design) (video)](https://www.youtube.com/watch?v=TMuno5RZNeE)
+        - [ ] [Thiết kế mẫu trong C# (SOLID Design Patterns in C#) (video)](https://www.youtube.com/playlist?list=PL8m4NUhTQU48oiGCSgCP1FiJEcg_xJzyQ)
```
Comment (dangkhoasdc) [inline README-vn.md:1162]: Mẫu thiết kế hướng đối tượng

File: README-vn.md [hunk 6]
```diff
@@ -1130,14 +1130,218 @@ Bạn sẽ biết thêm nhiều ứng dụng của đồ thị trong sách của
...
+        - [ ] [Đệ qui đuôi là gì Tại sao nó không tốt(What Is Tail Recursion Why Is It So Bad)?](https://www.quora.com/What-is-tail-recursion-Why-is-it-so-bad)
+        - [ ] [Đệ qui đuôi (Tail Recursion) (video)](https://www.youtube.com/watch?v=L1jjXGfxozc)
+
+- ### Lập trình hướng đối tượng
+    - [ ] [Phụ thêm: Seri về UML 2.0 (Optional: UML 2.0 Series) (video)](https://www.youtube.com/watch?v=OkC7HKtiZC0&list=PLGLfVvz_LVvQ5G-LdJ8RLqe-ndo7QITYc)
+    - [ ] Lập trình hướng đối tượng: Phát triển phần mềm sử dụng UML và Java (21 videos):
+        - Có thể bỏ qua phần này nếu bạn đã có kiến thức vững chắc về OOP
+        - [Lập trình hướng đối tượng: Phát triển phần mềm sử dụng UML và Java](https://www.youtube.com/playlist?list=PLJ9pm_Rc9HesnkwKlal_buSIHA-jTZMpO)
+    - [ ] Những điểm quan trọng nhất của OOP:
+        - [ ] [Bob Martin Những điểm quan trọng nhất của lập trình hướng đối tượng và thiết kế linh hoạt (Bob Martin SOLID Principles of Object Oriented and Agile Design) (video)](https://www.youtube.com/watch?v=TMuno5RZNeE)
```
Comment (dangkhoasdc) [inline README-vn.md:1161]: SOLID principle should be: nguyên tắc SOLID

File: README-vn.md [hunk 7]
```diff
@@ -1130,14 +1130,218 @@ Bạn sẽ biết thêm nhiều ứng dụng của đồ thị trong sách của
...
+        - [ ] [Mỗi đối tượng chịu một trách nhiệm duy nhất (Single responsibility to each Object)](http://www.javacodegeeks.com/2011/11/solid-single-responsibility-principle.html)
+            - [Đọc thêm](https://docs.google.com/open?id=0ByOwmqah_nuGNHEtcU5OekdDMkk)
+        - [ ] O - [Nguyên tắc mở đóng (Open/Closed Principal)](http://www.oodesign.com/open-close-principle.html)  | [Trong thực tế, đối tượng có thể mở rộng chứ không thích hợp cho sửa đổi (On production level Objects are ready for extension for not for modification)](https://en.wikipedia.org/wiki/Open/closed_principle)
+            - [Đọc thêm](http://docs.google.com/a/cleancoder.com/viewer?a=v&pid=explorer&chrome=true&srcid=0BwhCYaYDn8EgN2M5MTkwM2EtNWFkZC00ZTI3LWFjZTUtNTFhZGZiYmUzODc1&hl=en)
+        - [ ] L - [Nguyên tắc thay thế của Liskov (Liskov Substitution Principal)](http://www.oodesign.com/liskov-s-substitution-principle.html) | [Nguyên tắc về lớp cơ bản và lớp dẫn xuất (Base Class and Derived class follow ‘IS A’ principal)](http://stackoverflow.com/questions/56860/what-is-the-liskov-substitution-principle)
```
Comment (dangkhoasdc) [inline README-vn.md:1169]: Nguyên tắc 'IS-A': mối quan hệ giữa lớp cơ sở và lớp kế thừa.

File: README-vn.md [hunk 8]
```diff
@@ -1130,14 +1130,218 @@ Bạn sẽ biết thêm nhiều ứng dụng của đồ thị trong sách của
...
+        - [ ] state
+        - [ ] memento
+        - [ ] iterator
+        - [ ] composite
+        - [ ] flyweight
+    - [ ] [Chapter 6 (Part 1) - Patterns (video)](https://youtu.be/LAP2A80Ajrg?list=PLJ9pm_Rc9HesnkwKlal_buSIHA-jTZMpO&t=3344)
+    - [ ] [Chapter 6 (Part 2) - Abstraction-Occurrence, General Hierarchy, Player-Role, Singleton, Observer, Delegation (video)](https://www.youtube.com/watch?v=U8-PGsjvZc4&index=12&list=PLJ9pm_Rc9HesnkwKlal_buSIHA-jTZMpO)
+    - [ ] [Chapter 6 (Part 3) - Adapter, Facade, Immutable, Read-Only Interface, Proxy (video)](https://www.youtube.com/watch?v=7sduBHuex4c&index=13&list=PLJ9pm_Rc9HesnkwKlal_buSIHA-jTZMpO)
+    - [ ] [Series of videos (27 videos)](https://www.youtube.com/playlist?list=PLF206E906175C7E07)
+    - [ ] [Bắt đầu với thiết kế mẫu(Head First Design Patterns)](https://www.amazon.com/Head-First-Design-Patterns-Freeman/dp/0596007124)
```
Comment (dangkhoasdc) [inline README-vn.md:1202]: Dẫn nhập mẫu thiết kế hướng đối tượng

File: README-vn.md [hunk 9]
```diff
@@ -1130,14 +1130,218 @@ Bạn sẽ biết thêm nhiều ứng dụng của đồ thị trong sách của
...
+        - Tôi biết có một cuốn sách kinh điển là "Design Patterns: Elements of Reusable Object-Oriented Software" nhưng cuốn Bắt đầu với thiết kế mẫu(Head First Design Patterns là một lựa chọn tuyệt vời cho người mới bắt đầu với OO.
+    - [ ] [Ghi nhớ: 101 Thiết kế mẫu và lưu ý dành cho lập trình viên(Handy reference: 101 Design Patterns & Tips for Developers)](https://sourcemaking.com/design-patterns-and-tips)
```
Comment (dangkhoasdc) [inline README-vn.md:1204]: mẫu thiết kế not Thiết kế mẫu

File: README-vn.md [hunk 10]
```diff
@@ -1130,14 +1130,218 @@ Bạn sẽ biết thêm nhiều ứng dụng của đồ thị trong sách của
...
+    - [ ] [Xác suất (Make School: Probability) (video)](https://www.youtube.com/watch?v=sZkAAk9Wwa4)
+    - [ ] [Xác suất và chuỗi Markov (Make School: More Probability and Markov Chains) (video)](https://www.youtube.com/watch?v=dNaJg-mLobQ)
+    - [ ] Khan Academy:
+        - Các khóa học:
+            - [ ] [Lý thuyết xác suất(Basic Theoretical Probability)](https://www.khanacademy.org/math/probability/probability-and-combinatorics-topic)
+        - 41 video ngắn và đơn giản:
+            - [ ] [Lý giải về xác suất (Probability Explained) (video)](https://www.youtube.com/watch?v=uzkc-qNVoOk&list=PLC58778F28211FA19)
+
+- ### NP, NP-Complete và thuật toán xấp xỉ gần đúng
+    - Cần biết về các trường hợp nổi tiếng của NP-complete như chuyên du hành doanh nhân hay vấn đề túi đựng đồ nghề(traveling salesman and the knapsack problem) để có thể trả lời khi người phỏng vấn hỏi một cách ẩn ý.
```
Comment (dangkhoasdc) [inline README-vn.md:1217]: 1. [bài toán người bán hàng](https://vi.wikipedia.org/wiki/B%C3%A0i_to%C3%A1n_ng%C6%B0%E1%BB%9Di_b%C3%A1n_h%C3%A0ng)
2. [Bài toán xếp ba lô](https://vi.wikipedia.org/wiki/B%C3%A0i_to%C3%A1n_x%E1%BA%BFp_ba_l%C3%B4)

Comment (thovo) [reply README-vn.md:1217]: You should check that before I made the commit :smile:

Comment (dangkhoasdc) [reply README-vn.md:1217]: I haven't noticed =))

Comment (thovo) [reply README-vn.md:1217]: Please ensure you reviewed carefully! I will make another commit

**Comment by thovo** [comment]
Can we go back to work now @letientai299, @dangkhoasdc ?

**Comment by letientai299** [comment]
I'm in a job hunting. Can't work. But I'm will happily merge your PRs if both of you have agreed on the translation. 

I'll back to this project ASAP.

**Comment by thovo** [comment]
Good luck with your search! I will translate the next part soon!

**Comment by thovo** [comment]
Can you merge now @letientai299 ?

**Pull request merged**